### PR TITLE
WIP: Make docker pull durable and reduce page cache pressure

### PR DESCRIPTION
**Pull request by petrosagg**
Front logo [Front conversations](https://app.frontapp.com/open/top_whh)

### CODE REVIEW

**Review by zozo123 — COMMENTED** [review]
LGTM

File: pkg/archive/archive.go [hunk 1]
```diff
@@ -460,6 +460,20 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 			return err
 		}
 	}
+
+	if hdr.Typeflag == tar.TypeReg || hdr.Typeflag == tar.TypeRegA {
+		file, err := os.Open(path)
+		if err != nil {
+			file.Close()
```
Comment (willnewton) [inline pkg/archive/archive.go:468]: If err is not nil then the file was not opened, so does not need to be closed in the error case. It also looks safe to do the close with a "defer file.Close()" once we know the file was successfully opened.

File: pkg/archive/archive.go [hunk 2]
```diff
@@ -745,6 +765,21 @@ loop:
 			return err
 		}
 	}
+
+	for relPath, _ := range dirtyDirs {
+		path := filepath.Join(dest, relPath)
+
+		dir, err := os.OpenFile(path, syscall.O_DIRECTORY, 0)
+		if err != nil {
+			dir.Close()
```
Comment (willnewton) [inline pkg/archive/archive.go:779]: As above I think it would be safe to omit the dir.Close() here.

File: pkg/archive/archive.go (hunk 1 above)
Comment (zozo123) [reply pkg/archive/archive.go:468]: @willnewton, Actually I prefer it as it's done in the original PR. Doing `defer` in a file traversal procedure which has recursive nature could be very tricky. As you you know there is a LIFO queue that accumulates all the defer jobs, those are executed only after you leave the function,  meanwhile you can easily exceed the ulimit of concurrently opened file descriptors.

Comment (zozo123) [reply pkg/archive/archive.go:468]: RE this part you are right,

```
file, err := os.Open(path)
		if err != nil {
			file.Close()
```

Should be done only if `file` isn't null.

**Pull request closed without merging**